boards: imx93_evk: add board.c to control board mux

CkovMk opened the pull request:
i.MX 93 EVK has a series of digital muxes controlled by ADP5585 GPIO expander. This commit adds board level initialization to such muxes to enable selecting mux using Kconfig option.

CkovMk commented:
this will be updated once #76195 is approved. DNM

JiafeiPan commented on boards/nxp/imx93_evk/Kconfig:
```diff
@@ -0,0 +1,13 @@
+config BOARD_MIMX93_EVK_A55_EXP_SEL_INIT_PRIO
+	int "i.MX 93 EVK board mux control init priority"
+	default 50
+	depends on DT_HAS_IMX93EVK_EXP_SEL_ENABLED
```
remove `depends on DT_HAS_IMX93EVK_EXP_SEL_ENABLED`

JiafeiPan commented on boards/nxp/imx93_evk/Kconfig:
```diff
@@ -0,0 +1,13 @@
+# Copyright 2024 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_MIMX93_EVK_A55_EXP_SEL_INIT_PRIO
```
Remove "A55" as it can be reused by M33 Core.

JiafeiPan commented on boards/nxp/imx93_evk/Kconfig:
```diff
@@ -0,0 +1,13 @@
+	default 50
+	depends on DT_HAS_IMX93EVK_EXP_SEL_ENABLED
+
+module = BOARD_MIMX93_EVK_A55
```
module = MIMX93_EVK_EXP_SEL ? at least remove A55.

JiafeiPan commented on boards/nxp/imx93_evk/board.c:
```diff
@@ -0,0 +1,55 @@
+/**
```
rename board.c to board_exp_sel.c ? as this file is a module for select exp_sel?

CkovMk replied:
`board.c` seems to be a conventional name for board level init, and may contain code for multiple dts node. See `boards/nordic/nrf9160dk/board.c` for an example. Let's keep it as-is.

CkovMk replied:
`board.c` seems to be a conventional name for board level init, and may contain code for multiple dts node. See `boards/nordic/nrf9160dk/board.c` for an example. Let's keep it as-is.

JiafeiPan commented on boards/nxp/imx93_evk/board.c:
```diff
@@ -0,0 +1,55 @@
+
+LOG_MODULE_REGISTER(board_control, CONFIG_BOARD_MIMX93_EVK_A55_LOG_LEVEL);
+
+#if DT_HAS_COMPAT_STATUS_OKAY(imx93evk_exp_sel)
```
if no imx93evk-exp-sel device node available,  we could also call board_init_exp_sel to set exp_sel to a default value, such as 'A", otherwise, exp_sel can be configured by dts node property

CkovMk replied:
To avoid pulling unnecessary code to project (and some basic zephyr tests) that dose not use EXP_SEL, I've added `CONFIG_BOARD_MIMX93_EVK_EXP_SEL_INIT` option.

JiafeiPan commented on boards/nxp/imx93_evk/board.c:
```diff
@@ -0,0 +1,55 @@
+	}
+	LOG_INF("EXP_SEL mux %c with priority %d",
+		pin_state ? 'B' : 'A',
+		CONFIG_BOARD_MIMX93_EVK_A55_EXP_SEL_INIT_PRIO
```
why print init priority? less log is better.

CkovMk replied:
I'd like to make it easier to check the init priorities. For example, flexcan2 has no dependency of board mux in the devicetree. An inverted init priority config would not trigger a build error. In such case It's hard to tell why flexcan init failed.

JiafeiPan reviewed (changes requested):
in general, looks fine for me, but need some update.

JiafeiPan commented:
> this will be updated once #76195 is approved. DNM

Why?

CkovMk commented:
Note that we HAVE TO override  board mux selection by other enabled Kconfig options. Otherwise, samples / tests will fail.

JiafeiPan commented on boards/nxp/imx93_evk/Kconfig:
```diff
@@ -0,0 +1,16 @@
+
+config BOARD_MIMX93_EVK_EXP_SEL_INIT
+	bool "Configure i.MX 93 EVK board mux control during init"
+	default n
```
```
default y
depends on GPIO_ADP5585
```
So that if GPIO ADP5585 is not enabled, this item will not be enabled.

CkovMk replied:
Current implimentation is the other way around:

if BOARD_MIMX93_EVK_EXP_SEL_INIT then enable GPIO_ADP5585. See Kconfig.defconfig .  Should we change this?

CkovMk committed:
boards: imx93_evk: add board.c to control board muxes

i.MX 93 EVK has a series of digital MUXes controlled by ADP5585 GPIO
expander. This commit adds board level initialization to such MUXes to
control these MUXes.

Signed-off-by: Chekhov Ma <chekhov.ma@nxp.com>

JiafeiPan commented:
@CkovMk , would you please update board document to add the information for how to configure exp_sel?

JiafeiPan commented on boards/nxp/imx93_evk/doc/index.rst:
```diff
@@ -85,6 +85,23 @@ Serial Port
+Note that the MUX option is automatically overriden if certain board function is
+enabled. For instance, if ``CONFIG_CAN`` is enabled, MUX A is selected even if
+``mux="B";`` is configured in dts, and an warning would be printed to the log.
+
```
Add dts node example:

	board_exp_sel: board-exp-sel {
		compatible = "imx93evk-exp-sel";
		mux-gpios = <&gpio_exp0 4 GPIO_ACTIVE_HIGH>;
		mux = "A";
	};

JiafeiPan commented on boards/nxp/imx93_evk/doc/index.rst:
```diff
@@ -85,6 +85,23 @@ Serial Port
+Following steps are required to configure the ``EXP_SEL`` signal:
+
+1. Enable Kconfig option ``CONFIG_BOARD_MIMX93_EVK_EXP_SEL_INIT``.
+2. Select ``mux="A";`` or ``mux="B";`` in ``&board_exp_sel`` devicetree node.
```
Add dts node example?
board_exp_sel: board-exp-sel {
	compatible = "imx93evk-exp-sel";
	mux-gpios = <&gpio_exp0 4 GPIO_ACTIVE_HIGH>;
	mux = "A";
};

JiafeiPan commented on boards/nxp/imx93_evk/doc/index.rst:
```diff
@@ -85,6 +85,23 @@ Serial Port
+This board configuration uses a series of digital multiplexers to switch between
+different board functions. The multiplexers are controled by a GPIO signal called
+``EXP_SEL`` from onboard GPIO expander ADP5585. This signal is configured if a
+board function that requires configuring the mux is enabled.
```
"EXP_SEL" can be configured to select function set "A" or "B" by dts configuration if board control moduel is enabled.

CkovMk commented:
updated based on comments and try fix ci compliance wrn.

CkovMk committed:
boards: imx93_evk: document the configuration of board muxes

Document the added board level driver to control EXP_SEL muxes

Signed-off-by: Chekhov Ma <chekhov.ma@nxp.com>
